src/main/java/htsjdk/samtools/SamReader.java

```diff
@@ -547,42 +547,27 @@ static AssertingIterator of(final CloseableIterator<SAMRecord> iterator) {
         }
 
         private final CloseableIterator<SAMRecord> wrappedIterator;
-        private SAMRecord previous = null;
-        private SAMRecordComparator comparator = null;
+        private SAMSortOrderChecker checker = null;
 
         public AssertingIterator(final CloseableIterator<SAMRecord> iterator) {
             wrappedIterator = iterator;
         }
 
         public SAMRecordIterator assertSorted(final SAMFileHeader.SortOrder sortOrder) {
-
-            if (sortOrder == null || sortOrder == SAMFileHeader.SortOrder.unsorted) {
-                comparator = null;
-                return this;
-            }
-
-            comparator = sortOrder.getComparatorInstance();
+            checker = new SAMSortOrderChecker(sortOrder);
             return this;
         }
 
         public SAMRecord next() {
             final SAMRecord result = wrappedIterator.next();
-            if (comparator != null) {
-                if (previous != null) {
-                    if (comparator.fileOrderCompare(previous, result) > 0) {
-                        throw new IllegalStateException(MessageFormat.format(
-                                "Records {0} ({1}:{2}) should come after {3} ({4}:{5}) when sorting with {6}",
-                                previous.getReadName(),
-                                previous.getReferenceName(),
-                                previous.getAlignmentStart(),
-                                result.getReadName(),
-                                result.getReferenceName(),
-                                result.getAlignmentStart(),
-                                comparator.getClass().getName())
-                        );
-                    }
+            if (checker != null) {
+                final SAMRecord previous = checker.getPreviousRecord();
+                if (!checker.isSorted(result)) {
+                    throw new IllegalStateException(String.format(
+                            "Record %s should come after %s when sorting with %s ordering.",
+                            previous.getSAMString().trim(),
+                            result.getSAMString().trim(), checker.getSortOrder()));
```

Comment:
I'd say either we dump all the relevant information, i.e. the `getSAMString`, or we use use the `getSortKey` which is customized to show only the relevant information.  We can push that into the enum to require the people adding new sort orders specify which information is relevant.  Not that people add many new sort orders....

It seems like more of an intrusive change to modify SamRecord.toString... but I can do whichever.